staging/src/k8s.io/component-base/kflag/kflag.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kflag
+
+import (
+	"github.com/spf13/pflag"
+	utilflag "k8s.io/apiserver/pkg/util/flag"
+)
+
+type FlagSet struct {
+	fs *pflag.FlagSet
+}
+
+// For now, from a migration standpoint, it probably makes sense to re-use existing pflag flagsets,
+// since components may register flags against them in multiple places and need to incrementally
+// migrate, or do other custom things like set the normalize func. In the future, we should
+// probably be constructing the internal pflag flagset on behalf of the user, with a common
+// normalize func, etc.
+func NewFlagSet(fs *pflag.FlagSet) *FlagSet {
+	return &FlagSet{
+		fs: fs,
+	}
+}
+
+func (fs *FlagSet) Parse(args []string) {
+	fs.fs.Parse(args)
+}
+
+// Basic Int32Var approach, including a helper so components don't
+// have to repeatedly implement the basic apply for int32s.
+
+type Int32Value struct {
+	name  string
+	value int32
+	fs    *pflag.FlagSet
+}
+
+func (fs *FlagSet) Int32Var(name string, def int32, usage string) *Int32Value {
+	v := &Int32Value{
+		name: name,
+		fs:   fs.fs,
+	}
+	fs.fs.Int32Var(&v.value, name, def, usage)
+	return v
+}
+
+// Replace the int32 altogether
+func (v *Int32Value) Replace(target *int32) {
+	if v.fs.Changed(v.name) {
+		*target = v.value
+	}
+}
+
+// Custom user-defined behavior
+func (v *Int32Value) Custom(apply func(value int32)) {
+	if v.fs.Changed(v.name) {
+		apply(v.value)
+	}
+}
+
+// Example of a more complicated structure. In this case, we include
+// two helpers, one to override the map completely, and another to
+// merge the map while respecting flag precedence.
+
+type MapStringBoolValue struct {
+	name  string
+	value map[string]bool
+	fs    *pflag.FlagSet
+}
+
+func (fs *FlagSet) MapStringBoolVar(name string, def map[string]bool, usage string) *MapStringBoolValue {
+	val := &MapStringBoolValue{
+		name:  name,
+		value: make(map[string]bool),
+		fs:    fs.fs,
+	}
+	for k, v := range def {
+		val.value[k] = v
+	}
+	fs.fs.Var(utilflag.NewMapStringBool(&val.value), name, usage)
+	return val
+}
+
+// Replace the map altogether
+func (v *MapStringBoolValue) Replace(target *map[string]bool) {
```

Comment:
I suggested `ApplyIfChanged` above, but obviously here `Replace`-vs-`Merge` has some sense. Still `Replace` is somehow strange for values. Actually it is a setter, no? Maybe just `Set`?

Reply:
I'd be ok with `Set`

Reply:
`Set`, `Merge`, `Apply` – looks like a good triple of verbs.